include/seastar/json/json_elements.hh

```diff
@@ -367,6 +377,47 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
     };
 }
 
+/*!
+ * \brief consume values from a coroutine generator \c gen and encode the items using the \c fun mapping function
+ * onto the \c out output_stream as a json array.
+ */
+template<Jsonable T>
+future<> generate_array_element(output_stream<char>& out, coroutine::experimental::generator<T>& gen) {
```

Comment:
I think we should accept it by value since we consume it.

Reply:
Also, the generator not consumed by generate_array_element.
The latter is called in an async loop by the function returned by `generate_array`:
```c++
    return [&gen] (output_stream<char>&& s) mutable -> future<> {
        auto out = std::move(s);
        auto f = co_await coroutine::as_future(generate_array_element(out, gen));
        co_await out.close();
        f.get();
    };
```

Reply:
Can we return noncopyable_function?

Maybe precluded by another requirement, too bad.

Reply:
```c++
struct json_return_type {
    sstring _res;
    std::function<future<>(output_stream<char>&&)> _body_writer;
    json_return_type(std::function<future<>(output_stream<char>&&)>&& body_writer) : _body_writer(std::move(body_writer)) {
    }
```

Let's do this:
1. Add a type alias for `std::function<future<> (output_stream<char>&&)>`
2. Add a new API_LEVEL where this type alias is a noncopyable_function

Reply:
To reduce future churn, place this in the internal subnamespace.

Reply:
@avikivity do you think the api change should be part of this series or as a follow up?

Here, the returned function captures the generator by reference and the generator is kept alive by the caller so it's ok to return the coroutine lambda as a copyable function.

Reply:
It could be a follow up.

Reply:
However, do place internal functions in an internal subnamespace.